### DISCUSSION TITLE
Is it possible to re-evaluate the CSG graph

### DISCUSSION THREAD

**Discussion by rsaccon** [elalish/manifold#119] [Q&A]
Let's assume I  have a solid made out of many transformations and boolean operations. Now I want the final mesh of that solid in lowest acceptable precision for screen rendering but the exported output mesh has to be in accordance to required precision for 3D printing or CNC machining. And maybe for the screen rendering I want interactively change some transformation parameters or so. Now the question is: Do i need to recreate that solid from scratch for each such modification or is there a way to re-evaluate the CSG graph of that solid (and possibly saving some memory allocations).

**Comment by pca006132** [comment]
Not an expert with modeling and geometry, but here is my take on this:

For re-evaluation: I think it really depends on how you represent the CSG graph and whether the manifolds are mutable. If they are mutable, this would be pretty hard because you have to think about how to cache them. If we do something like the lazy boolean operations in #114, I think we can speed up re-evaluation of the CSG graph by storing the cached result and its hash, so when we need to re-evaluate the CSG graph, we can just propagate the changed objects and reuse the cached result for those that did not change. Not simple but probably easier than a mutable version.

For precision: I don't think you can increase the precision without recreating the entire solid with higher precision. Cache can only work if they have the same precision I guess. The smooth function may provide visually appealing results but probably not something you want to use for 3D printing/CNC machining.

**Reply by elalish** [reply]
Also, this Boolean library is overkill for rendering. You can do CSG in a pixel shader without computing any intersections at all if all you want is a visualization. Manifoldness doesn't matter there either.

**Reply by rsaccon** [reply]
OK, now I am trying to wrap my head around the idea of not using manifold for rendering. I have done before a bit of SDF based rendering, where all is done in the shader, but here we are talking about mesh based modeling and lets assume the render engine expects also a mesh. 

@elalish so you are saying I don't need to do any boolean operations for rendering ? If it is just for  unions this is easy (just a bit of depth ordering), but if it is for all boolean operations, then the shader needs to know, for each triangle, which boolean operation to apply. Do you know how this technique is called so I can find out more about it ?

**Reply by elalish** [reply]
Image based CSG, I guess? You'd probably just want to use this: http://opencsg.org/

**Reply by hrgdavor** [reply]
> Image based CSG, I guess? You'd probably just want to use this: http://opencsg.org/

I was looking around the web, and could not find. Are you aware of webgl variant of opencsg to use in JS ?

**Reply by elalish** [reply]
Not off the top of my head. But WebGL shaders are the same as OpenGL shaders, and I'd guess that's where most of the complexity is. I'd bet you could port openCSG to run in a three.js `CustomShader` without a crazy amount of effort.
